src/FSharp.Core/prim-types.fs

```diff
@@ -1654,27 +1654,148 @@ namespace Microsoft.FSharp.Core
                         else i <- i + 1
                     res
 
+            let isStructuralEquatable (ty:Type) = typeof<IStructuralEquatable>.IsAssignableFrom ty
+            let isArray (ty:Type) = ty.IsArray || (typeof<Array>.IsAssignableFrom ty)
 
-            /// One of the two unique instances of System.Collections.IEqualityComparer. Implements PER semantics
-            /// where equality on NaN returns "false".
-            let fsEqualityComparerNoHashingPER = 
-                { new IEqualityComparer with
-                    override iec.Equals(x:obj,y:obj) = GenericEqualityObj false iec (x,y)  // PER Semantics
-                    override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
-                    
-            /// One of the two unique instances of IEqualityComparer. Implements ER semantics
-            /// where equality on NaN returns "true".
-            let fsEqualityComparerNoHashingER = 
-                { new IEqualityComparer with
-                    override iec.Equals(x:obj,y:obj) = GenericEqualityObj true iec (x,y)  // ER Semantics
-                    override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
+            let canUseDefaultEqualityComparer er rootType =
+                let processed = HashSet ()
+
+                let rec checkType idx (types : Type array) =
```

Comment:
I would like to see more tests added in ways that really exercise all the different elements of logic here.

In particular I wonder if this function should be exposed in Debug build and its logic tested independently (it's so much easier to test that way).

Reply:
Alright so I added all the benchmarks for all the cases touched I think, or at least everything I could come up with :) the results are in the PR description.